oak_runtime/src/node/http/user.rs

```diff
@@ -0,0 +1,303 @@
+            privilege,
+        })
+    }
+
+    /// Creates a `request` and a `response` channel for communication with the Oak node. Uses the
+    /// `request` channel to send the HTTP request to the Oak node, and returns a handle to the
+    /// response channel for receiving the response from the Oak node.
+    fn inject_http_request(
+        &self,
+        runtime: RuntimeProxy,
+        outer_http_invocation: OuterHttpInvocation,
+    ) -> Result<oak_abi::Handle, ()> {
+        // Create a pair of temporary channels to pass the HTTP request to the Oak node, and to
+        // receive the response from it.
+        let pipe = Pipe::new(
+            &runtime,
+            &outer_http_invocation
+                .request_label
+                .ok_or(())
+                .map_err(|_err| {
+                    warn!("Cannot retrieve request_label from the OuterHttpInvocation");
+                })?,
+            &self.privilege,
+        )?;
+
+        // Put the HTTP request message inside the per-invocation request channel.
+        pipe.insert_message(
+            &runtime,
+            outer_http_invocation.request.ok_or(()).map_err(|_err| {
+                warn!("Cannot retrieve request from the OuterHttpInvocation");
+            })?,
+        )?;
+
+        // Send an invocation message (with attached handles) to the Oak Node.
+        trace!(
+            "UserNode {:?}: Sending invocation on outer_http_invocation.invocation_sender",
+            self.node_name
+        );
+        pipe.send_invocation(
+            &runtime,
+            outer_http_invocation
+                .invocation_sender
+                .ok_or(())
+                .map_err(|_err| {
+                    warn!("Cannot retrieve invocation_sender from the OuterHttpInvocation");
+                })?
+                .handle
+                .handle,
+        )?;
+
+        // Close all local handles except for the one that allows reading responses.
+        pipe.close(&runtime);
+
+        Ok(pipe.response_reader)
+    }
+
+    fn try_run(
+        self: Box<Self>,
+        runtime: RuntimeProxy,
+        startup_handle: oak_abi::Handle,
+    ) -> Result<(), ()> {
+        let receiver = Receiver::<OuterHttpInvocation>::new(ReadHandle {
+            handle: startup_handle,
+        });
+        let outer_http_invocation = receiver.receive(&runtime).map_err(|err| {
+            error!(
+                "{}: Failed to retrieve the outer HTTP invocation: {:?}",
+                self.node_name, err
+            )
+        })?;
+
+        // Send the request to the Oak node and get a handle to the channel containing the response.
+        let response_reader =
+            self.inject_http_request(runtime.clone(), outer_http_invocation.clone())?;
+        let response_receiver = crate::io::Receiver::new(ReadHandle {
+            handle: response_reader,
+        });
+        let response = response_receiver.receive(&runtime).map_err(|err| {
+            error!(
```

Comment:
Optional: I personally don't like this pattern of using `map_err` to do logging. I think it is cleaner if mapper functions like `map_err` do not produce side-effects. One option might be to define an error type that contains the message and it's severity. This can then be logged as eitehr a warning or error at the higher level where the result is processed.

Reply:
Will `anyhow::Context` fit here?